src/ipc/handler.c

```diff
@@ -671,7 +671,10 @@ static int ipc_pm_gate(uint32_t header)
 
 	IPC_COPY_CMD(pm_gate, _ipc->comp_data);
 
-	/* TODO: handle the flags */
+	if (pm_gate.flags & SOF_PM_PPG)
+		pm_runtime_disable(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID);
+	else
+		pm_runtime_enable(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID);
```

Comment:
Are we abusing the definition of 'prevent power gating'?
It looks from the commit message that you use this PPG flag for a slew of other tricks beyond just power gating.

Reply:
This is my understanding of pm-runtime api documented here https://www.kernel.org/doc/Documentation/power/runtime_pm.txt.

pm_runtime_enable/disable() is a simple ref counting to know how many clients wants to disable pm_runtime of a specific resource. In case of dsp master core, disabled power management means that core is locked in d0, does not go to d3 (its power is not gated) in waiti.

That is why PPG (Prevent Power Gating) set in flags disables pm_runtime, otherwise pm_runtime is enabled (by the host as a client in this case, there is another internal client = kpb).

Reply:
@mmaka1 this looks like a confusion: pm_runtime_enable/disable are not refcounted for the kernel. They do what they are supposed to do, enable or disable, and do so without context on how many times they were called.
The only ref counts are with pm_runttime_get/put, which will btw no have any effect unless pm_runtime_enable is called.

So if you wanted to do a refcount, you needed a pm_runtime_get()/put

Adding @ranj063 for a second opinion in case I need more coffee

Reply:
I attached a piece of the pm-runtime sof-docs I'm working on to illustrate the flows. 

As far as I understand the kernel docs and comments on the kernel mailing lists, both APIs are ref-counted. pn_runtime_enable/disable uses dev->power.disable_count and callbacks are run depending on the value of that ref-counter (reported by pm_runtime_active() /Liam req'ed me to rename this api to pm_runtime_is_active() in the fw/).